hydra/hydra.go

```diff
@@ -192,7 +205,73 @@ func NewHydra(ctx context.Context, options Options) (*Hydra, error) {
 		go handleBootstrapStatus(hdCtx, bsCh)
 
 		hds = append(hds, hd)
+
+		// dial back hooks
+		evts := []interface{}{
+			new(event.EvtPeerIdentificationCompleted),
+		}
+
+		subs, err := hd.Host.EventBus().Subscribe(evts, eventbus.BufSize(256))
+		if err != nil {
+			return nil, fmt.Errorf("head could not subscribe to eventbus events; err: %w", err)
+		}
+		go func(hd *head.Head, subs event.Subscription) {
+			defer subs.Close()
+			for {
+				select {
+				case evt := <-subs.Out():
+					ev := evt.(event.EvtPeerIdentificationCompleted)
+					p := ev.Peer
+					addrs := hd.Host.Peerstore().Addrs(p)
```

Comment:
Which addresses will the peer store contain? Only the address obtained via identify, or also (potentially local interface) addresses that the peer might have announced?

Reply:
@marten-seemann Both. 

We receive all address advertised by a peer on Identify protocol -> add them all to the peerstore . The addresses a peer listens on for it's local interfaces are also sent by it over the Identify protocol.

Reply:
That will lead to a lot of dial failures then which are unrelated to NATting, won't it?

Reply:
We consider a dial to be a failure here ONLY if we exhaust all addresses and still aren't able to dial a peer. That's how the Swarm dial logic works.

Reply:
Are we meant to measure if AutoNAT is working or how many of the nodes are actually dialable? If the latter then I'm concerned that if there are AutoNAT bugs then we'll run into issues because the list of addresses does not the peer's actual remote address. If it's the former then this seems reasonable.

Reply:
In this case, I think we want to know "is the peer dialable period".

Reply:
@aschmahmann It's the latter. What do you mean by the peer's actual remote address ? You mean the remote address we'd see on a connection to it ? That is usually discovered by the peer using AutoNAT post which it includes it in it' s set of dialable addresses and works pretty well right now.

Reply:
Yep, I mean the remote address the peer dials out from and that we see when they connect to us. I agree that if everything with AutoNAT is working fine and we don't connect to the hydra nodes before learning our addresses via AutoNAT then this should be good. However, if either of those conditions aren't true then we may get some false reports of NAT'd nodes.